scientific/src/commonMain/kotlin/formatter/CommonFormatter.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ Copyright 2022 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.scientific.formatter
+
+import com.splendo.kaluga.scientific.ScientificValue
+import com.splendo.kaluga.scientific.unit.ScientificUnit
+import kotlin.math.roundToLong
+
+private typealias CustomFormatHandler = (Number) -> String
+class CommonFormatter private constructor(builder: Builder) : Formatter {
+    class Builder private constructor() {
+        companion object {
+            fun build(build: Builder.() -> Unit = {}) : Formatter {
+                val builder = Builder()
+                build(builder)
+                return CommonFormatter(builder)
+            }
+        }
+
+        internal val customFormatters = mutableMapOf< ScientificUnit<*>, CustomFormatHandler>()
+        fun ifUnitIs(unit: ScientificUnit<*>, format: CustomFormatHandler) {
+            customFormatters[unit] = format
+        }
+    }
+
+    companion object {
+        val default = Builder.build()
```

Comment:
I think you can make class sealed, builder private and then define:
```
companion object Default : CommonFormatter(Builder())
```
Probably then you can omit private constructors...

Reply:
Not necessary against it but I don't think it is a better way to do it. What are the advantages of this approach?

About private constructors,  I used it to force users to use the build method, what is wrong with it? 

I like in the sealed class approach that we can add a new customised formatter as a subclass but don't like that we made the builder's constructor public. 02bce26903c748660566560373bad54f5b28b2dd

Reply:
Advantage is clear, no one can create CommonFormatter (as you did before with private constructor).
Extra feature having `companion object` allows you to do call like this:
```
CommonFormatter.format()
```
No need to do `CommonFormatter.Default.format()` because you already provide default object (by only one possible companion object)

Reply:
But it is exactly what I wanted: always use the builder for the common constructor of using predefined options

Reply:
But good point about `companion object`. Nice syntax sugar :D. 
* Made default formatted a companion object

Reply:
:+1: